chainerrl/agents/categorical_dqn.py

```diff
@@ -0,0 +1,138 @@
+        scatter_add = xp.scatter_add
+    else:
+        scatter_add = np.add.at
+
+    z_probs = xp.zeros((batch_size, n_atoms), dtype=xp.float32)
+    offset = xp.arange(
+        0, batch_size * n_atoms, n_atoms, dtype=xp.int32)[..., None]
+    # Accumulate m_l
+    scatter_add(
+        z_probs.ravel(),
+        (l.astype(xp.int32) + offset).ravel(),
+        (y_probs * (u - bj)).ravel())
+    # Accumulate m_u
+    scatter_add(
+        z_probs.ravel(),
+        (u.astype(xp.int32) + offset).ravel(),
+        (y_probs * (bj - l)).ravel())
+    # Deal with the case when bj is an integer, i.e., l = u = bj
+    scatter_add(
+        z_probs.ravel(),
+        (u.astype(xp.int32) + offset).ravel(),
+        (y_probs * (u == l)).ravel())
+    return z_probs
+
+
+class CategoricalDQN(dqn.DQN):
+    """Categorical DQN.
+
+    See https://arxiv.org/abs/1707.06887.
+
+    Arguments are the same as those of DQN except q_function must return
+    DistributionalDiscreteActionValue and clip_delta is ignored.
+    """
+
+    def _compute_target_values(self, exp_batch, gamma):
+        """Compute a batch of target return distributions."""
+
+        batch_next_state = exp_batch['next_state']
+        target_next_qout = self.target_model(batch_next_state)
+
+        batch_rewards = exp_batch['reward']
+        batch_terminal = exp_batch['is_state_terminal']
+
+        batch_size = exp_batch['reward'].shape[0]
+        z_values = target_next_qout.z_values
+        n_atoms = z_values.size
+
+        # next_q_max: (batch_size, n_atoms)
+        next_q_max = target_next_qout.max_as_distribution.data
+        assert next_q_max.shape == (batch_size, n_atoms), next_q_max.shape
+
+        # Tz: (batch_size, n_atoms)
+        Tz = (batch_rewards[..., None]
+              + (1.0 - batch_terminal[..., None]) * gamma * z_values[None])
+        return _apply_categorical_projection(Tz, next_q_max, z_values)
+
+    def _compute_y_and_t(self, exp_batch, gamma):
+        """Compute a batch of predicted/target return distributions."""
+
+        batch_size = exp_batch['reward'].shape[0]
+
+        # Compute Q-values for current states
+        batch_state = exp_batch['state']
+
+        # (batch_size, n_actions, n_atoms)
+        qout = self.model(batch_state)
+        n_atoms = qout.z_values.size
+
+        batch_actions = exp_batch['action']
+        batch_q = qout.evaluate_actions_as_distribution(batch_actions)
+        assert batch_q.shape == (batch_size, n_atoms)
+
+        with chainer.no_backprop_mode():
+            batch_q_target = self._compute_target_values(exp_batch, gamma)
+            assert batch_q_target.shape == (batch_size, n_atoms)
+
+        return batch_q, batch_q_target
+
+    def _compute_loss(self, exp_batch, gamma, errors_out=None):
+        """Compute a loss of categorical DQN."""
+        y, t = self._compute_y_and_t(exp_batch, gamma)
+        # minimize the cross entropy
+        eltwise_loss = -t * F.log(F.clip(y, 1e-10, 1.))
```

Comment:
Could you explain why `F.clip` is here?

Reply:
I found clipping was necessary for training CategoricalDQN from earlier experiments. Without clipping, some probability values converges to 0, resulting in log(0) -> NaN.

Other unofficial implementations also apply clipping.
https://github.com/Kaixhin/Rainbow/blob/master/agent.py#L85
https://github.com/floringogianu/categorical-dqn/blob/master/policy_improvement/categorical_update.py#L53

Since clipping by 1e-10 worked, I didn't tune 1e-10 further. It is possible larger values may result in better performance.

Reply:
I added a comment to explain why.